sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -813,6 +860,37 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 
+/**
+ * Unwrap the input of IsNull/IsNotNull if the input is NullIntolerant
+ * E.g. IsNull(Not(null)) == IsNull(null)
+ */
+object NullDownPropagation extends Rule[LogicalPlan] {
+  // Not all NullIntolerant can be propagated
+  // Return false if the expression may return null without non-null inputs.
+  // E.g. Cast is NullIntolerant; however, cast('Infinity' as integer) returns true
+  // Cannot apply to `ExtractValue` as the query planner uses the trait to resolve the columns.
+  // E.g. the planner may resolve column `a` to `a#123`, then IsNull(a#123) cannot be optimized
+  // Cannot apply to `EqualTo` as applying this optimization is too disruptive for some tests.
+  // E.g. [SPARK-32290]
+  private def supportedNullIntolerant(e: NullIntolerant): Boolean = (e match {
+    case _: Not => true
+    case _: GreaterThan | _: GreaterThanOrEqual | _: LessThan | _: LessThanOrEqual => true
```

Comment:
suppose we're converting `IsNull(a > b)` to `Or(IsNull(a), IsNull(b))`, what if `b` is non-deterministic? since its evaluation could be skipped.

Reply:
Thank you. Good point. Updated!